### PR TITLE
Addition to #234: Validate social fields

### PR DESCRIPTION
**Pull request by drepram**
I'm going to make this clear. I'm not going to claim that I did all of the parts in this PR, major changes were created by https://github.com/Suryansh5545. What I did was help out by cleaning out some logs, renamed some variables, and fixed the logic.

I visibly forked the repo from his account and not Cloud-CV's.

#### Changes proposed in this pull request:

There was a problem:

When you change the URL of a social field, the submit button wouldn't be triggered. You have to change a part of the identification field {first_name, last_name, affiliation}. 

I fixed this by adding conditions that would allow the submit button to be triggered, and at the same time also validating them using a simple DOM method.

- Add URL verification
- Fix bug where modification of social fields wouldn't trigger a submission
- Link to live demo: http://pr-243-evalai.surge.sh
- Screen Shot 2019-12-17 at 23 48 36 1

### CODE REVIEW

**Comment by Sanji515** [comment]
After entering the 2 valid url if I enter last url invalid then it shows URL invalid error message to all the three fields, this is happening as you are using `inputErrorMessage` to display message which is same for all the input. @drepram Please fix this.

**Comment by drepram** [comment]
Screen Shot 2019-12-18 at 22 42 15
<hr>
Screen Shot 2019-12-18 at 22 42 31
<hr>
I've created a local variable to store all of their errors together. I assume this is what you meant. Please take a look.

File: package.json [hunk 1]
```diff
@@ -44,7 +44,7 @@
     "zone.js": "^0.8.29"
   },
   "devDependencies": {
-    "@angular-devkit/build-angular": "~0.13.0",
+    "@angular-devkit/build-angular": "^0.12.4",
```
Comment (Sanji515) [inline package.json]: @drepram Why this change?

File: src/app/components/profile/profile.component.html [hunk 2]
```diff
@@ -1,148 +1,152 @@
...
+                  </li>
+                </ul>
+              </div>
+            </div>
+          </div>
+        </div>
+      </div>
+      <app-footer class="container-fluid"></app-footer>
+    </div>
+  </div>
```
Comment (Sanji515) [inline src/app/components/profile/profile.component.html]: Can you please revert unnecessary changes in this file?

File: src/app/components/utility/modal/modal.component.ts [hunk 3]
```diff
@@ -230,33 +235,75 @@ export class ModalComponent implements OnInit {
...
-        this.inputErrorMessage = 'Old password cannot be same as New Password';
-      }
-    } else if (e.target.name === 'new_password2') {
-      this.retype_newPassword = e.target.value;
-      if (e.target.value !== this.newPassword) {
-        this.inputErrorMessage = 'Password do not match';
-      }
-    }
+  validURL(string) { // https://stackoverflow.com/a/49849482/10103199
+    const res = string.match(/(http(s)?:\/\/.)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)/g);
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: I'm able to enter url without `http` or `https` due to which it gives backend error as these are `URL fields` on the backend. @drepram

File: src/app/components/utility/modal/modal.component.ts [hunk 4]
```diff
@@ -230,33 +235,75 @@ export class ModalComponent implements OnInit {
...
+     } else if (e.target.name === 'editChallengeTitle') {
+       this.isDisabled = e.target.value === this.challenge.title;
+     } else if (e.target.name === 'update_first_name') {
+       this.isDisabled = e.target.value === this.user.first_name;
+     } else if (e.target.name === 'update_last_name') {
+       this.isDisabled = e.target.value === this.user.last_name;
+     } else if (e.target.name === 'update_affiliation') {
+       this.isDisabled = e.target.value === this.user.affiliation;
+     } else if (e.target.name === 'update_google_scholar_url') {
+       if (this.validURL(e.target.value)) {
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: You should also add condition for empty, these url fields are not required so a user can update his profile without entering the url fields.

File: src/app/components/utility/modal/modal.component.ts [hunk 5]
```diff
@@ -230,33 +235,75 @@ export class ModalComponent implements OnInit {
...
+      if (this.validURL(e.target.value)) {
+        this.isDisabled = e.target.value === this.user.linkedin_url;
+
+        this.invalidFieldsAsText = this.invalidFields.length ? `${this.invalidFields.length > 1 ? `${this.invalidFields.join(' and ')} are` : `${this.invalidFields.join('')} is`} not valid` : null;
+        this.invalidFields = this.invalidFields.filter(element => element !== 'LinkedIn');
+       } else {
+        this.isDisabled = false; // In case they want to empty the field.
+
+        if (!(this.invalidFields.find(element => element === 'LinkedIn'))) { this.invalidFields.push('LinkedIn'); }
+        this.invalidFieldsAsText = this.invalidFields.length ? `${this.invalidFields.length > 1 ? `${this.invalidFields.join(' and ')} are` : `${this.invalidFields.join('')} is`} not valid` : null;
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: This `and` for three field doesn't look good, can we add `,` in between if it's length is 3?

File: yarn.lock [hunk 6]
```diff
@@ -0,0 +1,9049 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (Sanji515) [inline yarn.lock]: @drepram We don't need this change.

File: src/app/components/utility/modal/modal.component.ts [hunk 7]
```diff
@@ -230,33 +235,75 @@ export class ModalComponent implements OnInit {
...
+       if (this.validURL(e.target.value)) {
+         this.isDisabled = e.target.value === this.user.google_scholar_url;
+
+         this.invalidFields = this.invalidFields.filter(element => element !== 'Google Scholar');
+         this.invalidFieldsAsText = this.invalidFields.length ? `${this.invalidFields.length > 1 ? `${this.invalidFields.join(' and ')} are` : `${this.invalidFields.join('')} is`} not valid` : null;
+       } else {
+         this.isDisabled = false; // In case they want to empty the field.
+
+         if (!(this.invalidFields.find(element => element === 'Google Scholar'))) { this.invalidFields.push('Google Scholar'); }
+         this.invalidFieldsAsText = this.invalidFields.length ? `${this.invalidFields.length > 1 ? `${this.invalidFields.join(' and ')} are` : `${this.invalidFields.join('')} is`} not valid` : null;
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: @drepram Also can you please change the error message color to orange.

File: package.json (hunk 1 above)
Comment (drepram) [reply package.json]: It was an error that occurred on my machine. Take a look at https://stackoverflow.com/a/56624167/10103199.

Comment (Sanji515) [reply package.json]: is this(^) error showed due to some change of yours?

File: src/app/components/utility/modal/modal.component.ts (hunk 4 above)
Comment (drepram) [reply src/app/components/utility/modal/modal.component.ts]: I don't think there's a need for an empty condition? The user can still empty their field even though an error would pop up.

Comment (Sanji515) [reply src/app/components/utility/modal/modal.component.ts]: yeah but you need to hide the error for that empty field

File: package.json (hunk 1 above)
Comment (drepram) [reply package.json]: I never installed Angular CLI, I installed it and it didn't worked. Googled it and a lot of people had the same problem, shall I revert the change?

Comment (Sanji515) [reply package.json]: yes please

File: src/app/components/profile/profile.component.html [hunk 8]
```diff
@@ -104,7 +104,7 @@
 								</a>
 							</li>
 							<li>
-								<button (click)="updateUserDetails()" class="btn ev-btn-dark waves-effect waves-dark grad-btn grad-btn-dark fs-14">Update Profile</button>
+								<button (click)="updateUserDetails()" class="btn ev-btn-dark waves-effect waves-dark grad-btn grad-btn-dark fs-14">Update PP</button>
```
Comment (Sanji515) [inline src/app/components/profile/profile.component.html]: Please revert this change.

File: src/app/components/profile/profile.component.html [hunk 9]
```diff
@@ -145,4 +145,4 @@
 	</div>
 	<app-footer class="container-fluid" ></app-footer>
 </div>
-</div>
+</div>
```
Comment (Sanji515) [inline src/app/components/profile/profile.component.html]: Please add a new blank line here.

File: package.json (hunk 1 above)
Comment (drepram) [reply package.json]: But then it wouldn't work my machine? Wouldn't that be bad assuming other people would have the same problem?

My machine is a very generic machine: `MacBook Pro (13-inch, 2017, Two Thunderbolt 3 ports`.

File: src/app/components/utility/modal/modal.component.ts [hunk 10]
```diff
@@ -230,33 +235,80 @@ export class ModalComponent implements OnInit {
...
+      if (this.validProtocol(e.target.value) && this.validURL(e.target.value)) {
+        this.isDisabled = e.target.value === this.user.github_url;
+
+        this.invalidFields = this.invalidFields.filter(element => element !== 'GitHub');
+        this.invalidFieldsAsText = this.invalidFields.length ? `${this.invalidFields.length > 1 ? (this.invalidFields.length > 2 ? `${this.invalidFields.join(', ')} are` : `${this.invalidFields.join(' and ')} are`) : `${this.invalidFields.join('')} is`} not valid` : null;
+       } else {
+        this.isDisabled = false; // In case they want to empty the field.
+
+        if (!(this.invalidFields.find(element => element === 'GitHub'))) { this.invalidFields.push('GitHub'); }
+        this.invalidFieldsAsText = this.invalidFields.length ? `${this.invalidFields.length > 1 ? (this.invalidFields.length > 2 ? `${this.invalidFields.join(', ')} are` : `${this.invalidFields.join(' and ')} are`) : `${this.invalidFields.join('')} is`} not valid` : null;
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: This doesn't look good. Can we simplify this expression? or move it in a separate place and then call it from here.

File: package.json (hunk 1 above)
Comment (Sanji515) [reply package.json]: But this isn't related to this PR, if you feel the issue is valid and it is occurring on the master branch too then please open a separate PR for that fix.

Comment (drepram) [reply package.json]: Fair point.

File: package.json [hunk 11]
```diff
@@ -32,8 +32,8 @@
     "@types/moment-timezone": "^0.5.12",
     "angular-froala-wysiwyg": "^3.0.0-rc.2-1",
     "core-js": "^2.4.1",
-    "materialize-css": "^0.97.8",
     "hammerjs": "^2.0.8",
+    "materialize-css": "^0.97.8",
```
Comment (pushkalkatara) [inline package.json]: Please revert this change.

File: src/app/components/profile/profile.component.html [hunk 12]
```diff
@@ -85,11 +75,21 @@
 								{{user.google_scholar_url}}
 								</div>
 							    </div>
+                </li>
```
Comment (pushkalkatara) [inline src/app/components/profile/profile.component.html:78]: I am not able to understand this change, could you please explain.

Comment (drepram) [reply src/app/components/profile/profile.component.html:78]: The order should be Google Scholar, GitHub, and LinkedIn. The order was not correct in the HTML and so I moved the positions.

File: package.json [hunk 13]
```diff
@@ -67,4 +67,4 @@
     "tslint": "^5.9.1",
     "typescript": "~3.2.0"
   }
-}
+}
```
Comment (Sanji515) [inline package.json]: Please add a new line here.

**Review by Sanji515 — COMMENTED** [review]
Screenshot from 2019-12-19 22-43-48

@drepram Please fix this(^). It should not show error message for linked In url  field. And also when there is single error then instead of `<url field> are not valid`, it should show `<url feld> is not valid`.

**Comment by drepram** [comment]
1. image
> Fixed.
2. Screen Shot 2019-12-20 at 00 40 56
> I already wrote it before you requested it.

**Review by Sanji515 — COMMENTED** [review]
Screenshot from 2019-12-20 12-06-08

I'm unable to submit (I've modified the url fields), @drepram Please fix this.

File: src/app/components/utility/modal/modal.component.ts [hunk 14]
```diff
@@ -230,6 +235,36 @@ export class ModalComponent implements OnInit {
     this.denyCallback();
   }
 
+  convertFieldArrayIntoText(fields) {
+    let finalText = '';
+
+    if (fields.length) {
+      if (fields.length >= 1) {
+        if (fields.length > 2) {
+          finalText = `${fields.join(', ')} are not valid. Use HTTPS.`;
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: Can we please remove `Use HTTPS` as we are also allowing `HTTP`?

File: src/app/components/utility/modal/modal.component.ts [hunk 15]
```diff
@@ -230,6 +235,36 @@ export class ModalComponent implements OnInit {
...
+
+    this.invalidFieldsAsText = finalText;
+  }
+
+  validProtocol(string) {
+    const res = string.match(/^(http|https):\/\/./g);
+    return (res !== null);
+  }
+
+  validURL(string) { // https://stackoverflow.com/a/49849482/10103199
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: @drepram Can we please remove this comment from here?

File: src/app/components/utility/modal/modal.component.ts [hunk 16]
```diff
@@ -242,6 +277,57 @@ export class ModalComponent implements OnInit {
       this.isDisabled = e.target.value === this.user.last_name;
     } else if (e.target.name === 'update_affiliation') {
       this.isDisabled = e.target.value === this.user.affiliation;
+    } else if (e.target.name === 'update_google_scholar_url') {
+      if (this.validProtocol(e.target.value) && this.validURL(e.target.value)) {
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: Can we please combine these (`validProtocol` and `validURL`) two functions into one for simplicity?

File: src/app/components/utility/modal/modal.component.ts [hunk 17]
```diff
@@ -242,6 +277,57 @@ export class ModalComponent implements OnInit {
...
+        this.isDisabled = e.target.value === this.user.google_scholar_url;
+
+        this.invalidFields = this.invalidFields.filter(element => element !== 'Google Scholar');
+        this.convertFieldArrayIntoText(this.invalidFields);
+      } else if (e.target.value === '') {
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: And can we move this case to the above `if` condition as only `isDisabled` needs to change a/c to this case?

Comment (drepram) [reply src/app/components/utility/modal/modal.component.ts]: We can't. 

Assume the following code:
```javascript
if (e.target.value === '') {
  this.isDisabled = false;

  this.invalidFields = this.invalidFields.filter(element => element !== 'Google Scholar');
  this.convertFieldArrayIntoText(this.invalidFields);
}

if (this.validURL(e.target.value)) {
  this.isDisabled = e.target.value === this.user.google_scholar_url;

  this.invalidFields = this.invalidFields.filter(element => element !== 'Google Scholar');
  this.convertFieldArrayIntoText(this.invalidFields);
} else {
  if (!(this.invalidFields.find(element => element === 'Google Scholar'))) {
    this.invalidFields.push('Google Scholar');
  }
  this.convertFieldArrayIntoText(this.invalidFields);
}
```

When `e.target.value` is equal to `''`, the `this.invalidFields` array will be empty. But then, it would also be perceived as not a valid URL (`this.validURL`), thus it would go into the else statement, and it will be added back to the array. This would not happen if it was an `else if` function. Since the condition has already exist, it would not go into the `else` case.

**Comment by drepram** [comment]
image

> I'm sorry, but I can't reproduce your issue. I'm sorry to pull this card but...

image

File: src/app/components/utility/modal/modal.component.ts (hunk 17 above)
Comment (Sanji515) [reply src/app/components/utility/modal/modal.component.ts]: no no, sorry for mis typo, actually I mean to move this inside `if` condition and modify `isDisabled` accordingly.

Comment (drepram) [reply src/app/components/utility/modal/modal.component.ts]: Done, please take a look..

Comment (drepram) [reply src/app/components/utility/modal/modal.component.ts]: Done, please take a look.

File: src/app/components/utility/modal/modal.component.ts [hunk 18]
```diff
@@ -242,6 +272,42 @@ export class ModalComponent implements OnInit {
       this.isDisabled = e.target.value === this.user.last_name;
     } else if (e.target.name === 'update_affiliation') {
       this.isDisabled = e.target.value === this.user.affiliation;
+    } else if (e.target.name === 'update_google_scholar_url') {
+      if (this.validURL(e.target.value) || e.target.value === '') {
+        this.isDisabled = (e.target.value === '' ? false : e.target.value === this.user.google_scholar_url);
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: > > I'm sorry, but I can't reproduce your issue. 

It will fail, if suppose I've entered the different value for 2 fields but now I'm entering the same value for last URL field as `this.user.google_scholar_url` then it will disable the button but it should not happen. Please let me know if I'm not clear.

Comment (drepram) [reply src/app/components/utility/modal/modal.component.ts]: You know what, I'm just gonna set the isDisabled variable as false. It wouldn't be a problem if they submit a duplicate, right?

File: src/app/components/utility/modal/modal.component.ts [hunk 19]
```diff
@@ -230,6 +235,31 @@ export class ModalComponent implements OnInit {
     this.denyCallback();
   }
 
+  convertFieldArrayIntoText(fields) {
+    let finalText = '';
+
+    if (fields.length) {
+      if (fields.length >= 1) {
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: I think you should not include `=` here.

File: src/app/components/utility/modal/modal.component.ts [hunk 20]
```diff
@@ -242,6 +272,48 @@ export class ModalComponent implements OnInit {
       this.isDisabled = e.target.value === this.user.last_name;
     } else if (e.target.name === 'update_affiliation') {
       this.isDisabled = e.target.value === this.user.affiliation;
+    } else if (e.target.name === 'update_google_scholar_url') {
+      if (this.validURL(e.target.value) || e.target.value === '') {
+        this.isDisabled = false;
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts:271]: @drepram Similarly for `isDisabled`, it will enable the button if we entering the field correctly at the last even though the error message is showing for the entered invalid URL. So how about if we move this `isDisabled` and set to true at the last of `validateModalInput` function with the conditions like checking if `invalidFields` array is not empty and also covering other cases too if there are any.

File: src/app/components/utility/modal/modal.component.ts [hunk 21]
```diff
@@ -230,6 +235,31 @@ export class ModalComponent implements OnInit {
...
 
+  convertFieldArrayIntoText(fields) {
+    let finalText = '';
+
+    if (fields.length) {
+      if (fields.length >= 1) {
+        if (fields.length > 2) {
+          finalText = `${fields.join(', ')} are not valid. \n\n\nBe sure to include HTTP/HTTPS.`;
+        } else {
+          finalText = `${fields.join(' and ')} are not valid. \n\n\nBe sure to include HTTP/HTTPS.`;
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: Hey how about to keep it simple by adding `,` only instead of `and` for the case of length=2 and 3 differently? Like `Google Scholar, GitHub, LinkedIn are not valid. Be sure to include HTTP/HTTPS.` or `Google Scholar, GitHub are not valid. Be sure to include HTTP/HTTPS.`

**Comment by drepram** [comment]
Done, please take a look.

File: src/app/components/utility/modal/modal.component.ts [hunk 22]
```diff
@@ -230,6 +235,27 @@ export class ModalComponent implements OnInit {
     this.denyCallback();
   }
 
+  convertFieldArrayIntoText(fields) {
+    let finalText = '';
+
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: @drepram Can we please remove this extra line from here?

File: src/app/components/utility/modal/modal.component.ts [hunk 23]
```diff
@@ -230,6 +235,27 @@ export class ModalComponent implements OnInit {
...
+    if (fields.length) {
+      if (fields.length > 1) {
+        finalText = `${fields.join(', ')} are not valid. \n\n\nBe sure to include HTTP/HTTPS.`;
+      } else {
+        finalText = `${fields.join('')} is not valid. \n\n\nBe sure to include HTTP/HTTPS.`;
+      }
+    } else {
+      finalText = '';
+    }
+
```
Comment (Sanji515) [inline src/app/components/utility/modal/modal.component.ts]: Same here.

**Review by Sanji515 — COMMENTED** [review]
Minor changes. Rest LGTM!

**Comment by pushkalkatara** [comment]
Looks good to me!

**Review by Ram81 — APPROVED** [review]
LGTM 👍

**Review by Sanji515 — APPROVED** [review]
Good job @drepram :+1: This can be merged now :tada:

**Review by RishabhJain2018 — APPROVED** [review]
